File: rxgooglemap/src/main/java/com/sdoward/rxgooglemap/PolygonClickFunc.java
package com.sdoward.rxgooglemap;

import com.google.android.gms.maps.*;
import com.google.android.gms.maps.model.*;

import io.reactivex.Observable;
import io.reactivex.ObservableEmitter;
import io.reactivex.ObservableOnSubscribe;
import io.reactivex.functions.Function;

class PolygonClickFunc implements Function<GoogleMap, Observable<Polygon>> {

    @Override
    public Observable<Polygon> apply(final GoogleMap googleMap) {
        return Observable.create(new ObservableOnSubscribe<Polygon>() {
            @Override
            public void subscribe(final ObservableEmitter<Polygon> emitter) {
                googleMap.setOnPolygonClickListener(new GoogleMap.OnPolygonClickListener() {
                    @Override
                    public void onPolygonClick(Polygon polygon) {
                        emitter.onNext(polygon);
                    }
                });
            }
        });
    }

}

File: rxgooglemap/src/main/java/com/sdoward/rxgooglemap/PolygonClickFunc.kt
package com.sdoward.rxgooglemap

import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.model.Polygon
import io.reactivex.Observable
import io.reactivex.functions.Function

class PolygonClickFunc : Function<GoogleMap, Observable<Polygon>> {

    override fun apply(t: GoogleMap): Observable<Polygon> {
        return Observable.create { subscriber ->
            t.setOnPolygonClickListener {
                subscriber.onNext(it)
            }
        }
    }

}